ambry-router/src/main/java/com/github/ambry/router/ReplicateBlobOperation.java

```diff
@@ -0,0 +1,490 @@
+      }
+    }
+  }
+
+  /**
+   * Processes {@link ServerErrorCode} received from {@code replica}. This method maps a {@link ServerErrorCode}
+   * to a {@link RouterErrorCode}.
+   * @param serverErrorCode The ServerErrorCode received from the replica.
+   * @return the {@link RouterErrorCode} mapped from server error code.
+   */
+  private RouterErrorCode processServerError(ServerErrorCode serverErrorCode) {
+    switch (serverErrorCode) {
+      case Blob_Authorization_Failure:
+        return RouterErrorCode.BlobAuthorizationFailure;
+      case Blob_Expired:
+        return RouterErrorCode.BlobExpired;
+      case Blob_Not_Found:
+        return RouterErrorCode.BlobDoesNotExist;
+      case Disk_Unavailable:
+      case Replica_Unavailable:
+        return RouterErrorCode.AmbryUnavailable;
+      default:
+        return RouterErrorCode.UnexpectedInternalError;
+    }
+  }
+
+  /**
+   * Perform the necessary actions when a request to a replica fails.
+   * @param replicaId the {@link ReplicaId} associated with the failed response.
+   * @param exception the {@link RouterException} associated with the failed response.
+   */
+  private void onErrorResponse(ReplicaId replicaId, RouterException exception) {
+    onErrorResponse(replicaId, exception, true);
+  }
+
+  /**
+   * Perform the necessary actions when a request to a replica fails.
+   * @param replicaId the {@link ReplicaId} associated with the failed response.
+   * @param exception the {@link RouterException} associated with the failed response.
+   * @param updateDataNodeMetrics {@code true} if data node metrics should be updated. {@code false} otherwise.
+   */
+  private void onErrorResponse(ReplicaId replicaId, RouterException exception, boolean updateDataNodeMetrics) {
+    operationTracker.onResponse(replicaId,
+        TrackedRequestFinalState.fromRouterErrorCodeToFinalState(exception.getErrorCode()));
+    setOperationException(exception);
+    routerMetrics.routerRequestErrorCount.inc();
+    if (updateDataNodeMetrics) {
+      routerMetrics.getDataNodeBasedMetrics(replicaId.getDataNodeId()).replicateBlobRequestErrorCount.inc();
+    }
+  }
+
+  /**
+   * Completes the {@code ReplicateBlobOperation} if it is done.
+   */
+  private void checkAndMaybeComplete() {
+    // operationCompleted is true if Blob_Authorization_Failure was received.
+    if (operationTracker.isDone() || operationCompleted) {
+      if (operationTracker.hasSucceeded()) {
+        operationException.set(null);
+      } else if (operationTracker.maybeFailedDueToOfflineReplicas()) {
+        operationException.set(
+            new RouterException("ReplicateBlobOperation failed possibly because some replicas are unavailable",
+                RouterErrorCode.AmbryUnavailable));
+      } else if (operationTracker.hasFailedOnNotFound()) {
+        operationException.set(new RouterException("ReplicateBlobOperation failed because of BlobNotFound",
+            RouterErrorCode.BlobDoesNotExist));
+      }
```

Comment:
ReplicateBlobOperation should not care about the offline replica and should not care about the not found error.

Reply:
Just want to know the code a little bit more.
ReplicateBlob is possible to hit Not Found or unavailable as well. 
Although the operationTracker.operationException should have the corresponding error already, looks like to me, there is no harm to have these two conditions? Can you explain it a little bit more?